hwdef:Add Svehicle-E2-mini flight controllers (#33328)
ArduPilot/ardupilot: libraries/AP_HAL_ChibiOS/hwdef/SVehicle-E2-mini/README.md

# S-Vehicle E2-mini Flight Controller

The S-Vehicle E2-mini Flight Controller produced by [S-Vehicle](http://svehicle.cn).

![E2mini.png](E2mini.png)

## Features

- STM32H743IIK6 microcontroller
- 2 IMUs: Bosch BMI088, InvenSense ICM-42688P
- bmm150 magnetometer
- bmp388 barometers
- microSD card slot
- 1 ETH network interface
- 6 UARTs plus USB
- USB-TypeC port
- 14 PWM outputs
- PPM/SBus input, DSM/SBus input
- 3 I2C ports
- 2 CAN ports
- 100Mbps Ethernet port
- 1 power monitor inputs (XT30 port)

## Pinout

![E2mini-Top.png](E2mini-Top.png)
![E2mini-Top-pinouts.png](E2mini-TOP-pinouts.png)
![E2mini-pinouts-side.png](E2mini-pinouts-side.png)

## UART Mapping

- SERIAL0 -> USB (MAVLink2)
- SERIAL1 -> UART7 (TELEM1, MAVLink2)
- SERIAL2 -> UART5 (TELEM2, MAVLink2)
- SERIAL3 -> USART1 (GPS1/Compass/Safety Switch)
- SERIAL4 -> UART8 (GPS2/Compass)

> - SERIAL5 -> USART2 (TELEM3)

Reviewer: ```suggestion
- SERIAL5 -> USART2 (TELEM3, USER)
```

- SERIAL6 -> UART4 (User)

> - SERIAL7 -> USART3 (Debug)

Reviewer: ```suggestion
- SERIAL7 -> USART3 (Debug,USER)
```

> - SERIAL8 -> USART6 (Generally used for SBUS)

Reviewer: ```suggestion
- SERIAL8 -> USART6 (RX pin only, USER)
```

> All UARTs have full DMA capability.

Reviewer: ```suggestion
Except for USART6, all UARTS have full dma capability
```

## RC Input

> RC input is configured on the RCIN pin, at one end of the servo rail, marked PPM in the above diagram. All ArduPilot supported unidirectional RC protocols can be input here including PPM. For bi-directional or half-duplex protocols, such as CRSF/ELRS a full UART will have to be used. See [RC systems](https://ardupilot.org/copter/docs/common-rc-systems.html) for details on USRT setup for other protocols.

Reviewer: the image does not have it shown...nor are the PWM outputs labeled...and TELEMx is miss-spelled often on the image..

Reviewer: unaddressed

Author: I have added an image, E2mini-PWM-pinouts.png, specifically for annotating the PWM pins, and also corrected the spelling error in TELEMx

## PWM Output

> The E2mini flight controller supports up to 12 PWM outputs.

Reviewer: ```suggestion
The E2mini flight controller supports up to 12 PWM outputs. All outputs except PWM9 and PWM10 support DShot.
```

Reviewer: need to mention BDShot capable outputs

Author: None of the PWM pins on this board are driven, so all outputs except PWM9 and PWM10 support both DShot and BDShot.

The 14 PWM outputs are in 4 groups:

- PWM 1-4 in group1 (TIM5)
- PWM 5-8 in group2 (TIM4)
- PWM 9, 10 in group3 (TIM12)
- PWM 11,12 in group4 (TIM3)

Channels within the same group need to use the same output rate. If any channel in a group uses DShot then all channels in the group need to use DShot.

## GPIOs

The complete list of GPIOS is:

| Label | GPIO number |
|---|---|
| PWM(1) | 50 |
| PWM(2) | 51 |
| PWM(3) | 52 |
| PWM(4) | 53 |
| PWM(5) | 54 |
| PWM(6) | 55 |
| PWM(7) | 56 |
| PWM(8) | 57 |
| PWM(9) | 58 |
| PWM(10) | 59 |
| PWM(11) | 60 |
| PWM(12) | 61 |

> | LED_RED | 90 |

Reviewer: remove the following 4 lines...users dont need to know them....I would add that the Buzzer is on GPIO 77 though

| LED_GREEN | 91 |
| LED_BLUE | 92 |
| SP3_DRDY2 | 93 |

## Battery Monitoring

> The board has a XT30 port,it can receive power input ranging from 2 to 14s, and it integrates a step-down module internally,in this way, an external voltage reduction module can be avoided, and the board can read the input voltage through the internal ADC.

Reviewer: ```suggestion
The board has a XT30 port marked POWER, and it can accept voltage inputs ranging from 2 to 14s providing an internal step-down module for the autopilot board power, avoiding the need for an external power module. Default parameters for the battery voltage and current monitor are provided. Note: that only the current consumed by the autopilot and its peripherals is reported.  Separate monitoring of motor current via external ESCs is NOT provided. A DroneCAN current monitor could be used for this.
```
 question: does it supply the servo rail 5V power also?

Author: I have made the changes as you requested. 
The positive plate of the PWM port does not provide power. The positive terminal of all PWMs outputs the voltage that is supplied from the outside

## Compass

> The E2-mini has an bmm150 builtin compass, but due to interference the board is usually used with an external I2C or CAN compass as part of a GPS/Compass combination.

Reviewer: ```suggestion
The E2-mini has a BMM150 built-in compass, but due to interference, the board is usually used with an external I2C or CAN compass as part of a GPS/Compass combination.
```

Reviewer: you removed this entire section! why??

Author: This was my mistake, and I have added it now

## Analog Inputs

The E2-mini has 5 analog inputs.

> - ADC Pin9 -> Battery Voltage

Reviewer: Only the external pin analog inputs need to be mentioned...remove the internal ones

Reviewer: this pin is not external...should remove the section entirely

| SERIAL2 | UART5 | MAVLink2 | Yes | Yes |
| SERIAL3 | USART1 | GPS | Yes | Yes |
| SERIAL4 | UART8 | GPS | Yes | Yes |
| SERIAL5 | USART2 | MAVLink2 | Yes | Yes |
| SERIAL6 | UART4 | None | Yes | Yes |
| SERIAL7 | USART3 | None | Yes | Yes |
| SERIAL8 | USART6 | RCIN | n/a | Yes |

All configured UARTs have DMA capability. USART6 is RX only.

## RC Input

RC input is configured by default on the dedicated `RC` pad (PI5, RCININT), and accepts all ArduPilot supported unidirectional protocols. The pad labeled `SBUS` in the pinout is USART6 RX, which maps to SERIAL8. SERIAL8 is configured for RC input by default (`SERIAL8_PROTOCOL` = `23`), so an SBUS receiver connected to the `SBUS` pad works out of the box.

For bi-directional or half-duplex protocols such as CRSF/ELRS or SRXL2, a spare full UART must be used instead. SERIAL6 (UART4, User) is available for this. See [Radio Control Systems](https://ardupilot.org/copter/docs/common-rc-systems.html) for details on setting up UART-based RC protocols.

## PWM Output

The E2-mini flight controller supports 11 PWM outputs, labeled `M1`-`M8` and `A1`-`A3` on the board.

The outputs are in 4 groups:

> - M1-M4 in group1 (TIM5). All four support DShot.

Reviewer: **M1-M4 support bidirectional DShot now** — this push added `BIDIR` to `TIM5_CH4` and
`TIM5_CH2`, which covers both timer pairs, so all four outputs in group1 get BDShot. The
README still only claims plain DShot.

```suggestion
- M1-M4 in group1 (TIM5). All four support DShot and bidirectional DShot (BDShot).
```

Author: Okay, it's already changed

# S-Vehicle E2-mini Flight Controller

The S-Vehicle E2-mini Flight Controller produced by [S-Vehicle](http://svehicle.cn).

![E2mini.png](E2mini.png)

## Features

- STM32H743IIK6 microcontroller
- 2 IMUs: Bosch BMI088, InvenSense ICM-42688P
- bmm150 magnetometer
- bmp388 barometers
- microSD card slot
- 1 ETH network interface
- 6 UARTs plus USB
- USB-TypeC port
- 14 PWM outputs
- PPM/SBus input, DSM/SBus input
- 3 I2C ports
- 2 CAN ports
- 100Mbps Ethernet port
- 1 power monitor inputs (XT30 port)

## Pinout

![E2mini-Top.png](E2mini-Top.png)
![E2mini-Top-pinouts.png](E2mini-TOP-pinouts.png)
![E2mini-pinouts-side.png](E2mini-pinouts-side.png)  
![E2mini-PWM-pinouts.png](E2mini-PWM-pinouts.png) 
## UART Mapping

- SERIAL0 -> USB (MAVLink2)
- SERIAL1 -> UART7 (TELEM1, MAVLink2)
- SERIAL2 -> UART5 (TELEM2, MAVLink2)
- SERIAL3 -> USART1 (GPS1/Compass/Safety Switch)
- SERIAL4 -> UART8 (GPS2/Compass)
- SERIAL5 -> USART2 (TELEM3,USER)
- SERIAL6 -> UART4 (User)
- SERIAL7 -> USART3 (Debug,USER)
- SERIAL8 -> USART6 (RX pin only, USER)

Except for USART6, all UARTS have full dma capability

## RC Input

> SBUS (USART6) can be used as an RC Input.

Reviewer: ```suggestion
  RC input is configured by default on the dedicated RC pin, and accepts all ArduPilot supported unidirectional
  protocols (PPM, etc). SBUS can also be input via SERIAL8 (USART6, RX pin only) but SERAIL8_PROTOCOL must be set to "23".

  For bi-directional or half-duplex protocols such as CRSF/ELRS or SRXL2, a spare full UART must be used instead —
  SERIAL6 (UART4, User) is available for this. See [Radio Control
  Systems](https://ardupilot.org/copter/docs/common-rc-systems.html) for details on setting up UART-based RC protocols.
```

Author: I have made the modifications according to your suggestion.

## PWM Output

The E2mini flight controller supports up to 12 PWM outputs. All outputs except PWM9 and PWM10 support DShot and BDShot.

The 14 PWM outputs are in 4 groups:

- PWM 1-4 in group1 (TIM5)
- PWM 5-8 in group2 (TIM4)
- PWM 9, 10 in group3 (TIM12)
- PWM 11,12 in group4 (TIM3)

Channels within the same group need to use the same output rate. If any channel in a group uses DShot then all channels in the group need to use DShot.

## GPIOs

The complete list of GPIOS is:

| Label | GPIO number |
|---|---|
| PWM(1) | 50 |
| PWM(2) | 51 |
| PWM(3) | 52 |
| PWM(4) | 53 |
| PWM(5) | 54 |
| PWM(6) | 55 |
| PWM(7) | 56 |
| PWM(8) | 57 |
| PWM(9) | 58 |
| PWM(10) | 59 |
| PWM(11) | 60 |
| PWM(12) | 61 |

> 

Reviewer: add buzzer gpio

Author: Already added

## Battery Monitoring

> The board has a XT30 port marked POWER, and it can accept voltage inputs ranging from 2 to 14s providing an internal step-down module for the autopilot board power, avoiding the need for an external power module. Default parameters for the battery voltage and current monitor are provided. Note: that only the current consumed by the autopilot and its peripherals is reported.  Separate monitoring of motor current via external ESCs is NOT provided. A DroneCAN current monitor could be used for this.

Reviewer: provide the default parameters, see template

Author: Already provided

> 

Reviewer: need to add info on nARMED pin

Author: I have added detailed comments for the nARMED pin in both hwdef.dat and the documentation. Please take another look. Thanks！

## Compass

> The board has a XT30 port marked POWER, and it can accept voltage inputs ranging from 2 to 14s providing an internal step-down module for the autopilot board power, avoiding the need for an external power module. Default parameters for the battery voltage and current monitor are provided. Note: that only the current consumed by the autopilot and its peripherals is reported.  Separate monitoring of motor current via external ESCs is NOT provided. A DroneCAN current monitor could be used for this.

Reviewer: this is not a Compass section...it belongs in previous section....a Compass serction IS needed however

Author: Yes, I have revised it already
